aip/0165.md

```diff
@@ -0,0 +1,102 @@
+---
+aip:
+  id: 165
+  state: reviewing
+  created: 2019-09-18
+permalink: /165
+redirect_from:
+  - /0165
+---
+
+# Criteria-based delete
+
+Occasionally, an API may need to provide a mechanism to delete a large number
+of resources based on given query parameters, rather than requiring the
+individual resource name of the resources being deleted.
+
+This is a rare case, reserved for situations where users need to delete
+thousands or more resources at once, in which case the normal Batch Delete
+pattern ([AIP-235][]) becomes infeasible.
+
+## Guidance
+
+**Important:** Most APIs **should** use only Delete ([AIP-135][]) or Batch
+Delete ([AIP-235][]) for deleting resources, and **should not** implement
+deleting based on criteria. This is because deleting is generally irreversible
+and it is possible for a user to irrecovably lose data with a careless query.
+
+An API **may** implement a Purge method to permit deleting a large number of
+resources based on a filter string; however, this **should** only be done if
+the Batch Delete ([AIP-235][]) pattern is insufficient to achieve the use case:
+
+```proto
+rpc PurgeBooks(PurgeBooksRequest) returns (PurgeBooksResponse) {
+  option (google.api.http) = {
+    post: "/v1/{parent=publishers/*}/books:purge"
+    body: "*"
+  }
+}
+```
+
+- The HTTP verb **must** be `POST`, and the `body` **must** be `"*"`.
+- The `parent` field **should** be included in the URI. If the API wishes to
+  support deletion across multiple parents, it **should** accept the `-`
+  character consistent with [AIP-159][].
+
+### Request message
+
+Purge methods implement a common request message pattern:
+
+```proto
+message PurgeBooksRequest {
+  // The publisher to purge books from.
+  // To purge books across publishers, send "publishers/-".
+  string parent = 1 [
+    (google.api.field_behavior) = REQUIRED,
+    (google.api.resource_reference) = {
+      child_type: "library-example.googleapis.com/Book"
+    }];
+
+  // A filter matching the books to be purged.
+  string filter = 2 [(google.api.field_behavior) = REQUIRED];
+
+  // Actually perform the purge.
+  // If `force` is set to false, the method will return a sample of
+  // resource names that would be deleted.
+  bool force = 3;
+}
+```
+
+- The `filter` field **should** be required, and **must** follow the same
+  semantics as in List methods (see [AIP-132][]).
+- The `force` field **must** be included. If it is not set, the API **must**
+  return a count of the resources that would be deleted as well as a sample of
+  those resources, without actually performing the deletion.
+
+### Response message
+
+Purge methods implement a common response message pattern:
+
+```proto
+message PurgeBooksResponse {
+  // The number of books that this request deleted (or, if `force` is false,
+  // the number of books that will be deleted).
+  int32 purge_count = 1;
```

Comment:
Per our discussion: we shoudl clarify that this may be an estimate, and cite AIP-158's use of total_size for comparison.